### PR TITLE
Implement main toolbar

### PR DESCRIPTION
**Pull request by rchiodo**
Implement all the commands on the main toolbar
Change format of ICells to be based on Jupyter's ICodeCell format

### CODE REVIEW

File: src/client/datascience/history-react/cell.tsx [hunk 1]
```diff
@@ -115,13 +150,11 @@ export class Cell extends React.Component<ICellProps, ICellState> {
...
+        let newText = this.extractInputText();
         // Set our input text based on the new state
-        if (newState) {
-          newText = this.props.cell.input;
-        } else {
-          if (this.props.cell.input.length > 0) {
-            newText = this.props.cell.input.split('\n', 1)[0];
+        if (!newState) {
+          if (newText.length > 0) {
+            newText = newText.split('\n')[0];
```
Comment (IanMatthewHuff) [inline src/client/datascience/history-react/cell.tsx]: The (,1) on the split parameter means that we return back after finding the first split and don't bother to split any more. Was that removed on purpose here? We still only need the first split visually, we don't care about the rest at this point.

Comment (rchiodo) [reply src/client/datascience/history-react/cell.tsx]: No that was a typo. I had changed the props.cell.input to be different so I had to refactor some of this and I didn't compare to what it was before. I'll put it back

---
In reply to: [226004992](https://github.com/vscode-python-datascience/vscode-python/pull/25#discussion_r226004992) [](ancestors = 226004992)

File: src/client/datascience/history-react/MainPanel.tsx [hunk 2]
```diff
@@ -204,24 +259,94 @@ export class MainPanel extends React.Component<IMainPanelProps, IState> implemen
...
+
+    private canRedo = () => {
+        return this.state.redoStack.length > 0 ;
+    }
+
+    private canUndo = () => {
+        return this.state.undoStack.length > 0 ;
+    }
+
+    private addToUndoStack = () => {
```
Comment (IanMatthewHuff) [inline src/client/datascience/history-react/MainPanel.tsx]: We want to keep our oldest state (so we can always go back) but when we hit the stack limit here don't we want to start throwing out older states and still add in the new state? That allows for the most granular action on undo.

Comment (rchiodo) [reply src/client/datascience/history-react/MainPanel.tsx]: Ah that's true. It's just skipping it. I should instead push it off the end.

---
In reply to: [226012232](https://github.com/vscode-python-datascience/vscode-python/pull/25#discussion_r226012232) [](ancestors = 226012232)

**Pull request merged**